supabase/migrations/30_manual_billing.sql

```diff
@@ -0,0 +1,352 @@
+-- and save those reports to billing_historicals.
+create or replace function internal.freeze_billing_month(billed_month timestamptz)
+returns integer as $$
+declare
+    tenant_row record;
+    tenant_count integer = 0;
+begin
+    for tenant_row in select tenant as tenant_name from tenants loop
+        insert into internal.billing_historicals
+        select
+            report->>'billed_prefix' as tenant,
+            (report->>'billed_month')::timestamptz as billed_month,
+            report
+        from internal.billing_report_202308(tenant_row.tenant_name, billed_month) as report
+        on conflict do nothing;
+
+        -- INSERT statements set FOUND true if at least one row is affected, false if no row is affected.
+        if found then
+          tenant_count = tenant_count + 1;
+        end if;
+    end loop;
+    return tenant_count;
+end
+$$ language plpgsql volatile;
+-- End internal billing report gen
+
+create or replace view invoices_ext as
+with has_bypassrls as (
+  select exists(select 1 from pg_roles where rolname = current_role and rolbypassrls = true) as bypass
+),
+authorized_tenants as (
+  select tenants.tenant
+  from tenants
+  left join has_bypassrls on true
+  left join auth_roles('admin') on tenants.tenant ^@ auth_roles.role_prefix
+  where has_bypassrls.bypass or auth_roles.role_prefix is not null
+),
+historical_bills as (
+  select
+    date_trunc('month', (report->>'billed_month')::timestamptz)::date as date_start,
+    (date_trunc('month', (report->>'billed_month')::timestamptz) + interval '1 month' - interval '1 day')::date as date_end,
+    report->>'billed_prefix' as billed_prefix,
+    report->'line_items' as line_items,
+    report->'subtotal' as subtotal,
+    report as extra
+  from internal.billing_historicals
+  -- inner join should give only rows that match the join condition
+  inner join authorized_tenants on billing_historicals.tenant ^@ authorized_tenants.tenant
+),
+manual_bills as (
+  select
+    date_start,
+    date_end,
+    manual_bills.tenant as billed_prefix,
+    jsonb_build_array(
+      jsonb_build_object(
+        'description', manual_bills.description,
+        'count', 1,
+        'rate', manual_bills.usd_cents,
+        'subtotal', manual_bills.usd_cents
+      )
+    ) as line_items,
+    to_jsonb(usd_cents) as subtotal,
+    'null'::jsonb as extra
+  from internal.manual_bills
+  inner join authorized_tenants on manual_bills.tenant ^@ authorized_tenants.tenant
+),
+current_month as (
+  select
+    date_trunc('month', (report->>'billed_month')::timestamptz)::date as date_start,
+    (date_trunc('month', (report->>'billed_month')::timestamptz) + interval '1 month' - interval '1 day')::date as date_end,
+    report->>'billed_prefix' as billed_prefix,
+    report->'line_items' as line_items,
+    report->'subtotal' as subtotal,
+    report as extra
+  from authorized_tenants, internal.billing_report_202308(authorized_tenants.tenant, now()) as report
```

Comment:
Right, this does still have the discussed problem of invoices "disappearing" when the month rolls over but we haven't yet frozen the prior month's bills.

To do this comprehensively we'd need generate the series of months between the largest `internal.billing_historicals` month and `now()`, and generate reports for each of them.

This also makes it ambugious for the caller whether an invoice is finalized or not. Should we add a boolean to distinguish?

Reply:
Riiiiiiiiight, I forgot about that. So yeah, I think the right thing to do is what you say: fill in the missing months by generating the invoice. The view already have an `invoice_type` column, so I'm thinking that the right way to identify these invoices is to set their `invoice_type` to something like `preview` or `pending` or something like that.

Reply:
We'll also need to set a cutoff month for how far back we generate these. Probably something like the greater of August 2023 and the tenant's creation month

Reply:
can we consolidate `current_month` and `generated_prior_months` into a single CTE? Would it then makes sense to have InvoiceTypes of "final", "preview", and "manual"? Or is there something special about current-month that makes it different from a prior month that hasn't been finalized?

Reply:
Nope I think that grouping makes sense: the UI needs to know whether the invoice is "real" or not in order to know whether to show stripe invoice buttons, which this `final`/`preview` distinction covers. I'll make this change 👍